tests/31sync/09archived.pl

```diff
@@ -0,0 +1,193 @@
+test "Left rooms appear in the archived section of sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id );
+
+      matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_leave_room( $user, $room_id );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Newly left rooms appear in the archived section of incremental sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id, $next );
+
+     matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
```

Comment:
You can avoid the always-messy antipattern of capturing outer lexicals for later mutation here, by doing

    matrix_leave_room( $user, $room_id )
       ->then_done( $body->{next_batch} );

Causing the subsequent `then` to be able to receive

    my ( $next ) = @_;
    matrix_sync( $user, filter => $filter_id, since => $next );

That might seem a little neater, though it's swings and roundabouts either way.